cluster/nginx.go

```diff
@@ -294,31 +337,154 @@ func UpdateNginxConfiguration(ctx *Context) chan error {
-		}
-		nginxConfiguration.WriteString(`
+		nginxConfiguration.WriteString(serverBlock)
+	}
+	nginxConfiguration.WriteString(`
 			}
 		`)
+	return nginxConfiguration.String()
+}
 
-		configMap := corev1.ConfigMap{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: "nginx-configuration",
-			},
-			Data: map[string]string{
-				"nginx.conf": nginxConfiguration.String(),
-			},
+// getGlobalBucketNamespaceConfiguration build the nginx configuration for global bucket name space
+func getGlobalBucketNamespaceConfiguration() string {
+	// build a list of upstreams for each tenant
+	var tenantUpstreams bytes.Buffer
+	tenantsStream := streamTenantService(10)
+	for tenantRes := range tenantsStream {
+		if tenantRes.Error != nil {
+			log.Println(tenantRes.Error)
+			continue
 		}
-		_, err = clientset.CoreV1().ConfigMaps("default").Update(&configMap)
-		if err != nil {
-			panic(err.Error())
+
+		tUps := `
+			upstream %s {
+				server %s:%d;
+			}
+`
+		tenantUpstreams.WriteString(fmt.Sprintf(tUps, tenantRes.Tenant.ShortName, tenantRes.Service, tenantRes.Port))
+	}
+
+	// build a mapping of access keys to upstreams (by tenant short name)
+	var destinationMapping bytes.Buffer
+	accessTenantStream := streamAccessKeyToTenantServices()
+	for accessTenantResult := range accessTenantStream {
+		if accessTenantResult.Error != nil {
+			log.Println(accessTenantResult.Error)
+			continue
 		}
+		accessTenant := accessTenantResult.AccessKeyToTenantShortName
 
-		err = <-ReDeployNginxResolver(ctx)
-		if err != nil {
-			ch <- err
-			return
+		mapLine := `			"%s" "%s";
+`
+		destinationMapping.WriteString(fmt.Sprintf(mapLine, accessTenant.AccessKey, accessTenant.TenantShortName))
+	}
+
+	var nginxConfiguration bytes.Buffer
+	nginxConfiguration.WriteString(`
+			user nginx;
+			worker_processes auto;
+			error_log /dev/stdout debug;
+			pid /var/run/nginx.pid;
+
+			events {
+				worker_connections  1024;
+			}`)
+
+	nginxConfiguration.WriteString(fmt.Sprintf(`
+			http {
+			upstream portalproxy {
+				server portal-proxy:80;
+			}
+			upstream tenancy {
+				server portal-proxy:80;
+			}
+
+			%s
+		
+			map $http_authorization $access_destination {
+			default               "";
+				"~*Credential=(?<access_key>.*?)\/" "$access_key";
+			}
+		
+			# map to different upstream backends based on header
+			map $access_destination $pool {
+				"" "portalproxy";
+`, tenantUpstreams.String()))
+
+	nginxConfiguration.WriteString(destinationMapping.String())
+	appDomain := getS3Domain()
+	nginxConfiguration.WriteString(fmt.Sprintf(`
+			}
+
+
+			log_format  main  '$http_host - $remote_addr - $remote_user [$time_local] "$request" '
+																		'$status $body_bytes_sent "$http_referer" '
+																		'"$http_user_agent" "$http_x_forwarded_for"';
+				server {
+					access_log /var/log/nginx/access.log main;
+					location / {
+						 proxy_set_header Upgrade $http_upgrade;
+						 proxy_set_header Connection "upgrade";
+						 client_max_body_size 0;
```

Comment:
0 means no limit in the body size? if yes I think we should declare a maximum limit, even if this is big , to avoid potential DOS

Reply:
Btw we don't need to add connection upgrade here that seems wrong @dvaldivia 

```
server {
 listen 80;
 server_name example.com;

 # To allow special characters in headers
 ignore_invalid_headers off;
 # Allow any size file to be uploaded.
 # Set to a value such as 1000m; to restrict file size to a specific value
 client_max_body_size 0;
 # To disable buffering
 proxy_buffering off;

 location / {
   proxy_set_header X-Real-IP $remote_addr;
   proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
   proxy_set_header X-Forwarded-Proto $scheme;
   proxy_set_header Host $http_host;

   proxy_connect_timeout  300;
   # Default is HTTP/1, keepalive is only enabled in HTTP/1.1
   proxy_http_version 1.1;
   proxy_set_header Connection "";

   proxy_pass http://localhost:9000; # If you are using docker-compose this would be the hostname i.e. minio
   # Health Check endpoint might go here. See https://www.nginx.com/resources/wiki/modules/healthcheck/
   # /minio/health/live;
 }
}
```

Reply:
@harshavardhana this is for the portal, not for S3